solr-search/src/main/java/edu/unc/lib/dl/search/solr/service/ChildrenCountService.java

```diff
@@ -0,0 +1,228 @@
+        } else {
+            // Starting from a base query
+            solrQuery = baseQuery.getCopy();
+            // Make sure we aren't returning any normal results
+            solrQuery.setRows(0);
+            // Remove all facet fields so we are only getting ancestor path
+            if (solrQuery.getFacetFields() != null) {
+                for (String facetField : solrQuery.getFacetFields()) {
+                    solrQuery.removeFacetField(facetField);
+                }
+            }
+        }
+
+        // Algorithm for getting child counts for list of objects currently involves
+        // retrieving counts for every object in Solr and then matching
+        // the counts back to the metadata object using a binary search. We
+        // should reevaluate this as things scale, but was introduced as more
+        // efficient than hundreds of queries per request.
+        String ancestorPathField = solrField(ANCESTOR_PATH);
+        solrQuery.setFacetMinCount(1);
+        solrQuery.addFacetField(ancestorPathField);
+
+        solrQuery.add("f." + ancestorPathField + ".facet.limit", Integer.toString(Integer.MAX_VALUE));
+        // Sort by value rather than count so that earlier tiers will come first in case the result gets cut off
+        solrQuery.setFacetSort("index");
+
+        try {
+            QueryResponse queryResponse = this.executeQuery(solrQuery);
+            assignChildrenCounts(queryResponse.getFacetField(ancestorPathField), containers, countKey);
+        } catch (SolrServerException e) {
+            throw new SolrRuntimeException(e);
+        }
+    }
+
+    private SolrQuery createBaseQuery(AccessGroupSet principals) {
+        SolrQuery solrQuery = new SolrQuery();
+
+        // Add access restrictions to query
+        StringBuilder query = new StringBuilder("*:*");
+        restrictionUtil.add(query, principals);
+        solrQuery.setQuery(query.toString());
+
+        solrQuery.setStart(0);
+        solrQuery.setRows(0);
+        solrQuery.setFacet(true);
+
+        // Restrict counts to stand alone content objects
+        solrQuery.addFilterQuery(makeFilter(RESOURCE_TYPE,
+                asList(AdminUnit.name(), Collection.name(), Folder.name(), Work.name())));
+        return solrQuery;
+    }
+
+    /**
+     * Assigns children counts to container objects from ancestor path facet results based on matching search values
+     *
+     * @param facetField
+     * @param containerObjects
+     * @param countName
+     */
+    private void assignChildrenCounts(FacetField ancestorsField, List<BriefObjectMetadata> containerObjects,
+            String countName) {
+        if (ancestorsField.getValues() == null) {
+            return;
+        }
+
+        // Get the list of counts per pid
+        List<Count> counts = ancestorsField.getValues();
+        // Determine the most efficient algorithm for searching the counts
+        boolean binarySearch = counts.size() > 64;
```

Comment:
why 64?

Reply:
I can't actually remember where 64 came from. My recollection from when this was written was that when I was looking up the binary search method there was some recommendation not to use it for data sets below a certain point (maybe they said 64). But to be honest it may not be worth the readability hit having two searches.

When I was working on this refactoring I played around with a way shorter implementation that basically called `getChildrenCount` for each container individually instead of doing the one mega query and sorting out the counts. It was way less code, but I couldn't tell from my testing if there was a performance hit since I couldn't load enough items into solr to find out (embedded server would die about 120000 records). I opted not to change it if there wasn't a clear benefit since I assume I implemented more complex approach because of scaling issues. Once we integrate metrics tracking, I'm hoping to have a measurement for this method so we can track if its a problem.

Reply:
👍